### PR TITLE
feat(ticket#47): Leave Requests List UI implementation

### PR DESCRIPTION
**Pull request by nadirhussain786**
## Overview

This PR introduces the implementation of the `Leave Requests List` UI. I have used mock data to populate the interface. It includes an Action render component that displays different actions based on the logged-in user's role and the request status.

### Implementation

I have used the `TanStack` table to render the list of leave requests along with their associated actions in the UI.
This table contain these cells

- leaveType
- startDate
- EndDate
- status
- reasonForLeave
- requestedBy
- Actions

### Resources

image

### CODE REVIEW

File: src/app/leaves/components/LeaveActionRender.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+const loggedUser = {
+  name: "John Smith",
+  role: "member",
+};
+
+const LeaveTableActionRender = ({
+  leaveRequestStatus,
+}: {
+  leaveRequestStatus: string | undefined;
```
Comment (Monfernape) [inline src/app/leaves/components/LeaveActionRender.tsx:20]: This type seems to be wrong. I remember we have enum defined for this, right? Also, why would that be undefined?

File: src/app/leaves/components/LeaveActionRender.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+}) => {
+  const handleViewDetails = () => {
```
Comment (Monfernape) [inline src/app/leaves/components/LeaveActionRender.tsx:22]: Do we have the ticket to handle all these missing implementations?

File: src/app/leaves/components/LeaveActionRender.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      id: 4,
+      onClick: handleRejectedRequest,
+      icon: <AlertCircleIcon size={16} />,
+    },
+  ];
+
+  const actionMenu = useMemo(() => {
+    switch (loggedUser.role) {
+      case MemberRole.Member:
+        return leaveRequestStatus === "pending"
```
Comment (Monfernape) [inline src/app/leaves/components/LeaveActionRender.tsx:84]: Enum please.

File: src/app/leaves/components/LeaveActionRender.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  const viewInfo = [
+    {
+      title: "View Details",
+      id: 1,
+      onClick: handleViewDetails,
+      icon: <InfoIcon size={16} />,
+    },
+  ];
+
+  const StatusActions = [
```
Comment (Monfernape) [inline src/app/leaves/components/LeaveActionRender.tsx:66]: Variables should be camel case.

File: src/app/leaves/components/LeaveActionRender.tsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    },
+    {
+      title: "Delete",
+      id: 3,
+      onClick: handleDeleteRequest,
+      icon: <TrashIcon size={16} />,
+    },
+  ];
+
+  const viewInfo = [
```
Comment (Monfernape) [inline src/app/leaves/components/LeaveActionRender.tsx:57]: We're creating bunch of arrays but none of them is memoized.

File: src/app/leaves/components/LeaveActionRender.tsx (hunk 1 above)
Comment (Monfernape) [reply src/app/leaves/components/LeaveActionRender.tsx:20]: This won't be the only thing you'd be needing, you'd need id and other stuff as well.

File: src/app/leaves/components/LeavesRequestList.tsx [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    endDate: "2024-11-25",
+    status: "reject",
+    reasonForLeave: "Travel",
+    requestedBy: "Michael Lee",
+  },
+];
+
+export const LeavesRequestList = () => {
+
+  const columns: ColumnDef<LeaveRequestsTypes>[] = useMemo(
```
Comment (Monfernape) [inline src/app/leaves/components/LeavesRequestList.tsx:75]: Nice

File: src/app/leaves/components/LeavesRequestList.tsx [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    []
+  );
+
+  const table = useReactTable({
+    data: leaveRequests,
+    columns,
+    getCoreRowModel: getCoreRowModel(),
+  });
+
+  const groupedData = useGroupedData(leaveRequests, "shift");
```
Comment (Monfernape) [inline src/app/leaves/components/LeavesRequestList.tsx:130]: Why are we grouping this table?

File: src/app/leaves/components/LeavesRequestList.tsx [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+  const toNavigate = () => {
```
Comment (Monfernape) [inline src/app/leaves/components/LeavesRequestList.tsx:131]: This doens't look like a correct name. Function name should use verb. How about navigateToDetails or wherever this is going.

File: src/app/leaves/components/LeavesRequestList.tsx [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    alert("nativgation function trigger");
+  };
+
+  const StandardPageProps = {
```
Comment (Monfernape) [inline src/app/leaves/components/LeavesRequestList.tsx:135]: Memoization missing

File: src/types/index.tsx [hunk 10]
```diff
@@ -42,6 +42,36 @@ export type UserDetails = {
...
+  reasonForLeave: string;
+  status: string;
+}
+export interface MembersProps {
+  data: Member[];
+  success: boolean;
+  message: string;
+}
+
+export interface LeaveRequestsTypes {
```
Comment (Monfernape) [inline src/types/index.tsx:59]: We need these types to be localized. If we don't intend to reuse them, they should be close to consumer.

File: src/types/index.tsx [hunk 11]
```diff
@@ -42,6 +42,36 @@ export type UserDetails = {
   status: UserStatus;
 };
 
+export interface LeaveRequestData {
```
Comment (Monfernape) [inline src/types/index.tsx:45]: Not a fan of interfaces with word `Data` in them.

File: src/types/index.tsx [hunk 12]
```diff
@@ -42,6 +42,36 @@ export type UserDetails = {
...
+  memberId: string;
+  leaveType: string;
+  startDate: Date;
+  endDate: Date;
+  reasonForLeave: string;
```
Comment (Monfernape) [inline src/types/index.tsx:50]: Didn't we change this property?

Comment (Monfernape) [reply src/types/index.tsx:50]: @muhammad-bilal-ali-khokhar I think we're looking at a problem where one PR is pending review while other still points out to main, causing old changes to be part of the PR too.

File: src/common/StatusBadge/LeavesStatusBadge.tsx [hunk 13]
```diff
@@ -0,0 +1,34 @@
+import React from "react";
+import { Badge } from "@/components/ui/badge";
+import { LeavesRequestStatus } from "@/types";
+
+const statusClasses = {
```
Comment (Monfernape) [inline src/common/StatusBadge/LeavesStatusBadge.tsx:5]: Nice

File: src/app/leaves/page.tsx [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import React from "react";
+import { LeavesRequestList } from "./components/LeavesRequestList";
+import { PageLayout } from "../layout/PageLayout";
```
Comment (Monfernape) [inline src/app/leaves/page.tsx:3]: Why relative import?

File: src/app/leaves/components/LeaveActionRender.tsx [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+      title: "View Details",
+      id: 1,
+      onClick: handleViewDetails,
+      icon: <InfoIcon size={16} />,
+    },
+  ];
+
+  const StatusActions = [
+    {
+      title: "Approved",
```
Comment (Monfernape) [inline src/app/leaves/components/LeaveActionRender.tsx:68]: This should be called `Approve` because it hasn't been approved yet.

**Review by Monfernape — CHANGES_REQUESTED** [review]
Deletion or rejection actions should be used with red color.

File: src/types/index.tsx (hunk 12 above)
Comment (nadirhussain786) [reply src/types/index.tsx:50]: Sir @Monfernape , this will not happen on my side in the future.

File: src/types/index.tsx (hunk 11 above)
Comment (nadirhussain786) [reply src/types/index.tsx:45]: I have fixed this by renaming as `LeaveRequestDetails`

File: src/app/leaves/components/LeaveActionRender.tsx (hunk 4 above)
Comment (nadirhussain786) [reply src/app/leaves/components/LeaveActionRender.tsx:66]: Fixed

File: src/app/leaves/components/LeaveActionRender.tsx (hunk 2 above)
Comment (nadirhussain786) [reply src/app/leaves/components/LeaveActionRender.tsx:22]: We still haven't created any tickets regarding the implementation of these handles.

File: src/app/leaves/components/LeaveActionRender.tsx (hunk 1 above)
Comment (nadirhussain786) [reply src/app/leaves/components/LeaveActionRender.tsx:20]: I have addressed this comment by using an enum for the status type and also added a `leaveRequestId` prop, which will be used when we implement the handlers.

File: src/app/leaves/components/LeaveActionRender.tsx (hunk 3 above)
Comment (nadirhussain786) [reply src/app/leaves/components/LeaveActionRender.tsx:84]: Fixed

File: src/app/leaves/components/LeaveActionRender.tsx (hunk 5 above)
Comment (nadirhussain786) [reply src/app/leaves/components/LeaveActionRender.tsx]: I have memoized all arrays

**Review by Mubashar1009 — APPROVED** [review]

File: src/app/leaves/components/LeaveActionRender.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (nadirhussain786) [reply src/app/leaves/components/LeaveActionRender.tsx]: I have used enum to fix this issues

**Review by muhammad-bilal-ali-khokhar — APPROVED** [review]

**Review by Monfernape — APPROVED** [review]

**Pull request merged**